File: 프로그래머스/unrated/181916. 주사위 게임 3/주사위 게임 3.java
import java.util.stream.IntStream;

class Solution {
    public int solution(int a, int b, int c, int d) {
		int[] dice = new int[7];
		dice[a]++;
		dice[b]++;
		dice[c]++;
		dice[d]++;

		int answer = 0;
		int cnt = (int) IntStream.of(dice).filter(i -> i > 0).count();

		switch (cnt) {
			case 1 -> answer = 1111 * a;
			case 2 -> {
				int n = (int) IntStream.of(dice).filter(i -> i == 3).count();

				if(n == 1) { // 3, 1
					int p = IntStream.rangeClosed(0, 6).filter(i -> dice[i] == 3).findAny().orElse(0);
					int q = IntStream.rangeClosed(0, 6).filter(i -> dice[i] == 1).findAny().orElse(0);
					answer = (int) Math.pow(10 * p + q, 2);
				} else {
					int p = IntStream.rangeClosed(0, 6).filter(i -> dice[i] == 2).min().orElse(0);
					int q = IntStream.rangeClosed(0, 6).filter(i -> dice[i] == 2).max().orElse(0);
					answer = (p + q) * Math.abs(p - q);
				}
			}
			case 3 -> {
				int q = IntStream.rangeClosed(0, 6).filter(i -> dice[i] == 1).min().orElse(0);
				int r = IntStream.rangeClosed(0, 6).filter(i -> dice[i] == 1).max().orElse(0);
				answer = q * r;
			}
			case 4 -> answer = Math.min(Math.min(a, b), Math.min(c, d));
		}

		return answer;
	}
}

File: 프로그래머스/unrated/181916. 주사위 게임 3/주사위 게임 3.kt
import kotlin.math.abs

class Solution {
    fun solution(a: Int, b: Int, c: Int, d: Int): Int {
        val answer = hashMapOf<Int, Int>()
        answer[a] = answer[a]?.plus(1) ?: 1
        answer[b] = answer[b]?.plus(1) ?: 1
        answer[c] = answer[c]?.plus(1) ?: 1
        answer[d] = answer[d]?.plus(1) ?: 1

        return when (answer.size) {
            1 -> 1111 * a
            2 -> {
                val p = answer.entries.firstOrNull { it.value == 3 }
                if (p != null) {
                    val q = answer.entries.first { it.value != 3 }.key
                    (10 * p.key + q) * (10 * p.key + q)
                } else {
                    val (q, p2) = answer.keys.toList()
                    (p2 + q) * abs(p2 - q)
                }
            }
            3 -> {
                val (q, r) = answer.entries.filter { it.value != 2 }.map { it.key }
                q * r
            }
            else -> answer.keys.minOrNull() ?: 1
        }
    }
}